### PR TITLE
Add support for running operator tests in envtest

### PR DESCRIPTION
**Pull request by Porges**
You can now run the controller tests with a flag `-envtest` to use the `envtest` infrastructure. This means you do not need to spin up a kubernetes cluster or run the operator on its own container.

### CODE REVIEW

File: hack/generated/controllers/suite_test.go [hunk 1]
```diff
@@ -41,19 +55,101 @@ func (tc *ControllerTestContext) SharedResourceGroupOwner() genruntime.KnownReso
 	return genruntime.KnownResourceReference{Name: tc.SharedResourceGroup.Name}
 }
 
-func setup() error {
+func setupEnvTest() (*rest.Config, error) {
+	envtestContext = &EnvtestContext{
+		testenv: envtest.Environment{
+			CRDDirectoryPaths: []string{
+				"../config/crd/bases/valid",
```
Comment (Porges) [inline hack/generated/controllers/suite_test.go:62]: TODO: I made this 'valid' directory to put only the needed CRDs in as not all of them are able to be loaded (due to the Any-type stuff that @babbageclunk is working on…)

File: hack/generated/pkg/testcommon/wait.go [hunk 2]
```diff
@@ -18,49 +19,49 @@ func WaitFor(ctx context.Context, timeout time.Duration, check func(context.Cont
 	waitCtx, cancel := context.WithTimeout(ctx, timeout)
 	defer cancel()
 
-	done := false
-
-	for !done {
+	for {
```
Comment (Porges) [inline hack/generated/pkg/testcommon/wait.go:22]: This was reworked as it slept for an additional period in the success case.

File: hack/generated/controllers/suite_test.go [hunk 3]
```diff
@@ -31,6 +38,13 @@ const (
 )
 
 var testContext ControllerTestContext
+var envtestContext *EnvtestContext
+
+type EnvtestContext struct {
```
Comment (matthchr) [inline hack/generated/controllers/suite_test.go:43]: Is it `Envtest` or `EnvTest`? I guess `envtest` is the name of the module so it's all one word?

File: hack/generated/controllers/suite_test.go [hunk 4]
```diff
@@ -41,19 +55,101 @@ func (tc *ControllerTestContext) SharedResourceGroupOwner() genruntime.KnownReso
...
+		return nil, errors.Wrapf(err, "creating ARM client")
+	}
+
+	log.Print("Registering custom controllers")
+	errs := controllers.RegisterAll(
+		envtestContext.manager,
+		armClient,
+		controllers.KnownTypes,
+		klogr.New(),
+		controller.Options{})
```
Comment (Porges) [inline hack/generated/controllers/suite_test.go:102]: @matthchr should we be setting anything here?

File: hack/generated/pkg/testcommon/wait.go [hunk 5]
```diff
@@ -18,49 +19,49 @@ func WaitFor(ctx context.Context, timeout time.Duration, check func(context.Cont
...
 
-// testMainWrapper is a wrapper that can be called by TestMain so that we can use defer
 func SetupTeardownTestMain(
 	m *testing.M,
-	skipShortTests bool,
+	skipSlowTests bool,
 	setup func() error,
 	teardown func() error) int {
 
-	if skipShortTests {
```
Comment (matthchr) [inline hack/generated/pkg/testcommon/wait.go:48]: Thanks -- old name was super wrong =)

**Review by matthchr — COMMENTED** [review]
LGTM but since you have a TODO you're still working on will save signoff til after you're done with that.

File: hack/generated/controllers/suite_test.go (hunk 4 above)
Comment (matthchr) [reply hack/generated/controllers/suite_test.go:102]: Looks like standard usage is showing just `concurrency(1)`, so I'm thinking no?

**Review by matthchr — APPROVED** [review]

**Pull request merged**